packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1881,33 +1883,47 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
   Future<T> pushAndRemoveUntil<T extends Object>(Route<T> newRoute, RoutePredicate predicate) {
     assert(!_debugLocked);
     assert(() { _debugLocked = true; return true; }());
-    final List<Route<dynamic>> removedRoutes = <Route<dynamic>>[];
-    while (_history.isNotEmpty && !predicate(_history.last)) {
-      final Route<dynamic> removedRoute = _history.removeLast();
-      assert(removedRoute != null && removedRoute._navigator == this);
-      assert(removedRoute.overlayEntries.isNotEmpty);
-      removedRoutes.add(removedRoute);
-    }
+    final Route<dynamic> precedingRoute = _history.last;
     assert(newRoute._navigator == null);
     assert(newRoute.overlayEntries.isEmpty);
-    final Route<dynamic> oldRoute = _history.isNotEmpty ? _history.last : null;
     newRoute._navigator = this;
     newRoute.install(_currentOverlayEntry);
     _history.add(newRoute);
+
+    Route<dynamic> oldRoute;
+    int removalIndex = 0;
+
+    for (Route<dynamic> route in _history) {
```

Comment:
you can use [firstWhere](https://api.dartlang.org/stable/2.4.0/dart-core/Iterable/firstWhere.html)